### ISSUE TITLE
Runtime error mechanism

### ISSUE THREAD

**Issue by DarkDrek** [pvdrz/pijama_legacy#78]
Currently we have a few possible failing operations (in `native_bin_op` and `native_un_op`)
- Divide by 0
- Reminder/Modulo by 0
- Out of bound shift operations
- Negative of i128::MIN
- Possibly having more than `usize` variables

They will currently `panic` on the rust side. So I propose we implement a way to detect these cases and stop evaluation when we reach such an operation and display an error message to the user.

We would need to add a new error to `LangError` perhaps like this:
```
pub enum LangError<'a> {
    Ty(#[from] TyError),
    Parse(ParsingError<'a>),
    Evaluation(EvaluationError)
}
```
And change `Machine::evaluate` to return `LangResult<Term>` and also change all called methods to also return a `LangResult`.

We could also add a new `Primitive` to allow user to create their own runtime errors.

**Comment by DarkDrek** [comment]
On another note what should the behavior be for int operations that go outside the range available for `i128`?

**Comment by pvdrz** [comment]
>     * Divide by 0
> 
>     * Reminder/Modulo by 0
> 
>     * Out of bound shift operations
> 
>     * Negative of i128::MIN

I agree that we should have a documented behavior for such cases. Maybe we can use some of the functions [here](https://github.com/christianpoveda/pijama/pull/12) to handle overflowing and division by zero.

> 
>     * Possibly having more than `usize` variables
> 

This one doesn't worry me as much. Having more than `usize` variables would require having more bytes allocated that the possible number in the platform's architecture, which is impossible.

> We would need to add a new error to `LangError` perhaps like this:
> 
> ```
> pub enum LangError<'a> {
>     Ty(#[from] TyError),
>     Parse(ParsingError<'a>),
>     Evaluation(EvaluationError)
> }
> ```
> 
> And change `Machine::evaluate` to return `LangResult<Term>` and also change all called methods to also return a `LangResult`.

We used to have those in https://github.com/christianpoveda/pijama/pull/12. I removed them to boost performance. I'm not saying we shouldn't reintroduce them, we should because those are unhandled errors in our language and should be handled by us, not by Rust. I'm saying that we should be conscious that these changes will hurt performance.

> We could also add a new `Primitive` to allow user to create their own runtime errors.

I like it but I wouldn't know how to type it in our current type system. One would be tempted to use `Unit` but that would be too limiting. For example, this program
```rust
def fact(x: Int): Int do
  if x < 0 do
    panic("factorial of negatives is undefined")
  elif x == 0 do
    1
  else
    x * factorial(x - 1)
  end
end
```
Wouldn't type-check because two branches return `Int` and one `Unit`. We could add some notion of subtyping and introduce the [never type](https://doc.rust-lang.org/std/primitive.never.html) for panics but I'd rather improve the type system to have `Option`s and `Result`s first.

**Comment by pvdrz** [comment]
> On another note what should the behavior be for int operations that go outside the range available for `i128`?

This is a discussion we need to have. It would be nice if we had some sort of flags to disable enable "overflow checks". Something like `pijama --overflow` where programs don't crash if integers overflow and we just let the operations wrap.

Scripting languages tend to use [BigInt](https://en.wikipedia.org/wiki/Bigint) arithmetic but I'm not a big fan of it because performance is greatly reduced. In fact, I'd argue that we should drop from `i128` to `i64` because `i128` is not truly native, it is software implemented.

Maybe we could have fixed-size arithmetic as default (like now) and provide arbitary-size arithmetic if the user decides it's worth the performance cost. Something like
```elixir
x = 6 # This is an `i64`
y: BigInt = 6 # This is arbitrary precision
```

**Comment by pvdrz** [comment]
Maybe we could consider some form of custom panicking? https://crates.io/crates/lazy-panic

That'd let us have custom panic behavior and traces in the context of the language without introducing a `Result`.

**Comment by DarkDrek** [comment]
> I'd argue that we should drop from i128 to i64 because i128 is not truly native, it is software implemented.

I'm in favor of this change. I would do it outside of this issue and measure performance. Do you want a new issue for it for discussion?

We can delay the new `Primitve` and implement it when our type system becomes more flexible.

> Maybe we could have fixed-size arithmetic as default (like now) and provide arbitary-size arithmetic if the user decides it's worth the performance cost.

Do we want that as a language feature, as a user defined type or in the (upcoming) standard library?

> Something like pijama --overflow where programs don't crash if integers overflow and we just let the operations wrap.

That is similar to the rust behavior `overflow-checks = false     # Disable integer overflow checks.`

**Comment by pvdrz** [comment]
> I'm in favor of this change. I would do it outside of this issue and measure performance. Do you want a new issue for it for discussion?

Yes, I'll open one for it. I'm afraid that our benchmarks will broke in the change if we've been using too large integers.

> We can delay the new `Primitve` and implement it when our type system becomes more flexible.

It sounds good to me :)

> Do we want that as a language feature, as a user defined type or in the (upcoming) standard library?

I think we could offer it as a language feature? I'm not sure yet. I don't think anyone would implement their own BigInt arithmetic library if the language provided one. In contrast with collections where people could think they need different tradeoffs for their data structures. But I'm not sure if that's the case.
 
> That is similar to the rust behavior `overflow-checks = false # Disable integer overflow checks.`

Yes exactly! But we would enforce that as a flag of our compiler instead of relying that someone rebuilds Pijama with the flag in `rustc`.

**Comment by pvdrz** [comment]
Maybe we should think how to lower this to `LIR`? I'd like to avoid doing some sort of 
```rust
if overflow_checks {
    // Do checked operations
} else {
    // Do overflowing operations
}
```
In `Machine`. Maybe we could abstract away `native_bin_op` and `native_un_op` into an `Arithmetic` trait and then have two zst `CheckedArith` and `OverflowArith` implementing each one of the behaviors.

Then we could do `Machine` generic over some `A: Arithmetic` and just dispatch the correct `A` during startup?

**Comment by DarkDrek** [comment]
That should work maybe we can add the resulting error as a associated type and set it for the non checked variant to `! Never` to in theory get no performance penalty.

**Comment by pvdrz** [comment]
I'll start adding some scaffolding to receive flags and make our compiler more like a CLI.

Update: Before I forget, we should also decide what will be our runtime error mechanism :P
